apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -845,124 +665,201 @@ do_ensure_all_iterators_closed(_DSSessionID) ->
-        rank = Rank,
-        beginning = ItBegin
+new_batch({StreamKey, Srs0}, BatchSize, Session = #{s := S0}, ClientInfo) ->
+    SN1 = emqx_persistent_session_ds_state:get_seqno(?next(?QOS_1), S0),
+    SN2 = emqx_persistent_session_ds_state:get_seqno(?next(?QOS_2), S0),
+    Srs1 = Srs0#srs{
+        first_seqno_qos1 = SN1,
+        first_seqno_qos2 = SN2,
+        batch_size = 0,
+        last_seqno_qos1 = SN1,
+        last_seqno_qos2 = SN2
     },
-    mnesia:write(?SESSION_STREAM_TAB, DSStream, write),
-    DSStream.
-
-%% must be called inside a transaction
--spec session_drop_streams(id()) -> ok.
-session_drop_streams(DSSessionId) ->
-    mnesia:delete(?SESSION_STREAM_TAB, DSSessionId, write).
-
-%% must be called inside a transaction
--spec session_drop_pubranges(id()) -> ok.
-session_drop_pubranges(DSSessionId) ->
-    RangeIds = session_read_pubranges(DSSessionId, write),
-    lists:foreach(
-        fun(RangeId) ->
-            mnesia:delete(?SESSION_PUBRANGE_TAB, RangeId, write)
+    {Srs, Inflight} = enqueue_batch(false, BatchSize, Srs1, Session, ClientInfo),
+    S1 = emqx_persistent_session_ds_state:put_seqno(?next(?QOS_1), Srs#srs.last_seqno_qos1, S0),
+    S2 = emqx_persistent_session_ds_state:put_seqno(?next(?QOS_2), Srs#srs.last_seqno_qos2, S1),
+    S = emqx_persistent_session_ds_state:put_stream(StreamKey, Srs, S2),
+    Session#{s => S, inflight => Inflight}.
+
+enqueue_batch(IsReplay, BatchSize, Srs0, Session = #{inflight := Inflight0}, ClientInfo) ->
+    #srs{
+        it_begin = ItBegin0,
+        it_end = ItEnd0,
+        first_seqno_qos1 = FirstSeqnoQos1,
+        first_seqno_qos2 = FirstSeqnoQos2
+    } = Srs0,
+    ItBegin =
+        case IsReplay of
+            true -> ItBegin0;
+            false -> ItEnd0
         end,
-        RangeIds
-    ).
+    case emqx_ds:next(?PERSISTENT_MESSAGE_DB, ItBegin, BatchSize) of
+        {ok, ItEnd, Messages} ->
+            {Inflight, LastSeqnoQos1, LastSeqnoQos2} = process_batch(
+                IsReplay, Session, ClientInfo, FirstSeqnoQos1, FirstSeqnoQos2, Messages, Inflight0
+            ),
+            Srs = Srs0#srs{
+                it_begin = ItBegin,
+                it_end = ItEnd,
+                %% TODO: it should be possible to avoid calling
+                %% length here by diffing size of inflight before
+                %% and after inserting messages:
+                batch_size = length(Messages),
+                last_seqno_qos1 = LastSeqnoQos1,
+                last_seqno_qos2 = LastSeqnoQos2
+            },
+            {Srs, Inflight};
+        {ok, end_of_stream} ->
+            %% No new messages; just update the end iterator:
+            {Srs0#srs{it_begin = ItBegin, it_end = end_of_stream, batch_size = 0}, Inflight0};
+        {error, _} when not IsReplay ->
+            ?SLOG(info, #{msg => "failed_to_fetch_batch", iterator => ItBegin}),
+            {Srs0, Inflight0}
+    end.
+
+%% key_of_iter(#{3 := #{3 := #{5 := K}}}) ->
+%%     K.
```

Comment:
:scissors: ?

Reply:
This could be useful for debugging later.